### DISCUSSION TITLE
OIDC login is impossible

### DISCUSSION THREAD

**Discussion by mulder999** [outline/outline#7612] [Self Hosting] [answered]
We use Keycloak for Single Sign-On (SSO) with Outline/Wiki. While the integration generally functions correctly, we occasionally encounter a problem where users get stuck in an infinite logout loop despite a successful login attempt.

**Frequency**

Intermittent

**To Reproduce**

1. Log in to Outline/Wiki using Keycloak SSO.
2. Observe that the user is continuously bounced between `/` and `/home` pages

> REM: Clearing local storage and cookies does not stop bouncing

**Expected behavior**
Users should be able to log in successfully and access their account without being bounced back indefinitely
**OR** users should be presented a new login screen

**Screenshots**

Extract of the web console showing the exception at the origin of the infinite logout loop attempt:
![image](https://github.com/user-attachments/assets/3b78b49d-ee46-48d9-8fa2-6c20e81cfa49)

**Outline (please complete the following information):**
- Install: self hosted docker image
- Version: sha256:9919cfc5130cafe2c0ccd2ea31c7b9a311faad8e2243524670d1edb1bddf7040

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Browser: Vivaldi 6.6.3271.57
 - KeyCloak 25.04

**Comment by linear** [comment]
<p><a href="https://linear.app/outline/issue/OLN-622/oidc-infinite-logout-loop-on-some-occasion">OLN-622 OIDC infinite logout loop on some occasion</a></p>

**Comment by tommoor** [comment]
What version are you using?

**Comment by mulder999** [comment]
v0.79.1 but we had that behavior at least once with some prior release.
Changing to another browser or restarting keycloak does not solve the issue.
Last time the user concerned had to wait about 24h. Incredibly annoying.

**Comment by mulder999** [comment]
The problem is actually much more severe than initially described. Currently, no user can connect to the system, forcing us to roll back to version 0.78.0 until we can reproduce the issue in the development environment and find a fix.

After a OIDC login, we notice a few successful API calls (using DEBUG=http) followed by a 401 Unauthorized response, which triggers an automatic logout.

Potentially suspicious changes include: #7428, #7446, and the update labeled "Improved performance by increasing concurrency on worker."

While reviewing the code, we have a few questions regarding some design decisions:

- StateStore's role seems to disallow replay attack, it is not stateless nor deterministic, what mecanism prevents out of order parallels requests-response
- in APIClient, a single 401 error triggers a logout, is it intended behavior ?
- Could you provide more details about how WebSocket authentication is handled and how it interacts with the rest of the system?

**Comment by mulder999** [comment, marked as answer]
I am glad to report version 0.80.2 seems to solve this issue.
